internal/version/service.go

```diff
@@ -0,0 +1,163 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package version
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+
+	"storj.io/storj/internal/sync2"
+)
+
+// Config contains the necessary Information to check the Software Version
+type Config struct {
+	ServerAddress  string        `help:"server address to check its version against" default:"https://satellite.stefan-benten.de/version"`
+	RequestTimeout time.Duration `help:"Request timeout for version checks" default:"0h1m0s"`
+	CheckInterval  time.Duration `help:"Interval to check the version" default:"0h15m0s"`
+}
+
+// Service contains the information and variables to ensure the Software is up to date
+type Service struct {
+	config  *Config
+	info    *Info
+	service string
+
+	Loop *sync2.Cycle
+
+	checked chan struct{}
+	mu      sync.Mutex
+	allowed bool
+}
+
+// NewService creates a Version Check Client with default configuration
+func NewService(config *Config, info *Info, service string) (client *Service) {
+	return &Service{
+		config:  config,
+		info:    info,
+		service: service,
+		Loop:    sync2.NewCycle(config.CheckInterval),
+		checked: make(chan struct{}, 0),
+		allowed: false,
+	}
+}
+
+// Run logs the current version information
+func (srv *Service) Run(ctx context.Context) error {
+	firstCheck := true
+	return srv.Loop.Run(ctx, func(ctx context.Context) error {
+		var err error
+		allowed, err := srv.checkVersion(ctx)
+
+		srv.mu.Lock()
+		srv.allowed = allowed
+		srv.mu.Unlock()
+
+		if firstCheck {
+			close(srv.checked)
+			firstCheck = false
+			if !allowed {
+				zap.S().Fatal(errOldVersion)
+			}
+		}
+
+		if err != nil {
+			zap.S().Errorf("Failed to do periodic version check: ", err)
+		}
+
+		return nil
+	})
+}
+
+// IsUpToDate returns whether if the Service is allowed to operate or not
+func (srv *Service) IsUpToDate() bool {
+	<-srv.checked
+
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+
+	return srv.allowed
+}
+
+// CheckVersion checks if the client is running latest/allowed code
+func (srv *Service) checkVersion(ctx context.Context) (allowed bool, err error) {
+	defer mon.Task()(&ctx)(&err)
+	accepted, err := srv.queryVersionFromControlServer(ctx)
+	if err != nil {
+		return true, err
```

Comment:
okay so it's safe to do, but still surprising. my vote is to leave the determination of what to do in an error case to the caller